src/e2e/app.ts

```diff
@@ -34,1429 +22,738 @@ export interface BrowserDownload {
-      const document = await this.document();
-      const projectName = await document.getByTestId("project-name");
-      return projectName.getNodeText();
-    };
-    return waitFor(async () => {
-      const value = await text();
-      expect(value).toEqual(match);
-    }, defaultWaitForOptions);
-  }
-
-  async findActiveApiEntry(text: string, headingLevel: string): Promise<void> {
-    // We need to make sure it's actually visible as it's scroll-based navigation.
-    const document = await this.document();
-    return waitFor(async () => {
-      const items = await document.$$(headingLevel);
-      const headings: { text: string; visible: boolean }[] = await Promise.all(
-        items.map((e: ElementHandle<Element>) =>
-          e.evaluate((node) => {
-            const text = (node as HTMLElement).innerText;
-            const rect = (node as HTMLElement).getBoundingClientRect();
-            const visible = rect.top >= 0 && rect.bottom <= window.innerHeight;
-            return { text, visible };
-          })
-        )
-      );
-      const match = headings.find((info) => info.visible && info.text === text);
-      expect(match).toBeDefined();
-    }, defaultWaitForOptions);
-  }
-
-  async findDocumentationTopLevelHeading(
-    title: string,
-    description?: string
-  ): Promise<void> {
-    const document = await this.document();
-    await document.findByText(
-      title,
-      {
-        selector: "h2",
-      },
-      defaultWaitForOptions
-    );
-    if (description) {
-      await document.findByText(description);
-    }
+    await this.editorTextArea.click();
+    await this.page.keyboard.press(`${this.modifierKey}+A`);
   }
 
-  async selectDocumentationSection(name: string): Promise<void> {
-    const document = await this.document();
-    const button = await document.findByRole("button", {
-      name: `View ${name} documentation`,
-    });
-    await button.click();
-    await this.awaitAnimation();
+  async pasteInEditor() {
+    // Simulating keyboard press CTRL+V works in Playwright,
+    // but does not work in this case potentially due to
+    // CodeMirror pasting magic
+    const clipboardText: string = await this.page.evaluate(
+      "navigator.clipboard.readText()"
+    );
+    await this.editorTextArea.evaluate((el, text) => {
+      const clipboardData = new DataTransfer();
+      clipboardData.setData("text/plain", text);
+      const clipboardEvent = new ClipboardEvent("paste", { clipboardData });
+      el.dispatchEvent(clipboardEvent);
+    }, clipboardText);
   }
 
-  async awaitAnimation(): Promise<void> {
-    const page = await this.page;
-    await page.waitForTimeout(300);
+  async typeInEditor(text: string): Promise<void> {
+    const textWithoutLastChar = text.slice(0, text.length - 3);
+    await this.editorTextArea.fill(textWithoutLastChar);
+    // Last few characters are typed separately to trigger editor suggestions
+    const [a, b, c] = text.slice(-3);
```

Comment:
Have refactored `typeInEditor` here: https://github.com/microbit-foundation/python-editor-v3/pull/1153/commits/2339e62c825de3547e28fda53b9894786f886bd0

I changed it to slice the last 2 characters instead of last 3 since that seems to be sufficient for making the autocomplete tests pass

From observation, the delays don't feel significant. The delays were also added to reduce flakiness. I have updated the comment to mention this: https://github.com/microbit-foundation/python-editor-v3/pull/1153